added evaluation script for PPHumanSeg model

labeeb-7z opened the pull request:
This PR is meant to add an evaluation script for the PP Human segmentation model present in the zoo as mentioned in this issue #119 .

I referred to the val script used by PaddleSeg which can be found [here](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.1/contrib/HumanSeg/val.py) and [here](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.1/paddleseg/core/val.py) and implemented the evaluation script for PPHumanSeg in the method other evaluation scripts on zoo are written.

This is the ouput I'm getting after running the evaluation script : 

![Screenshot from 2023-02-10 09-55-30](https://user-images.githubusercontent.com/94868003/218001722-527e24ef-3d12-4c3b-89df-cd55f935b329.png)

Let me know if this PR can be improved further. Thanks!

fengyuentau commented on tools/eval/README.md:
```diff
@@ -181,7 +182,7 @@ python eval.py -m crnn -d icdar -dr /path/to/icdar
 ### Prepare data
 
-Please visit https://github.com/cv-small-snails/Text-Recognition-Material to download the IIIT5K dataset and the labels.
+Please visit https://github.com/cv-small-snails/Text-Recognition-Material to download the IIIT5K dataset and the labels. 
```
Extra whitespace

fengyuentau commented on tools/eval/README.md:
````diff
@@ -190,4 +191,18 @@ Run evaluation with the following command:
+```
+
+
+## PPHumanSeg
````
I wonder if we can use the `mini_supervisely` for the name of dataset so as to avoid confusion with the model itself.

fengyuentau commented on tools/eval/datasets/pp_humanseg.py:
```diff
@@ -0,0 +1,209 @@
+from tqdm import tqdm
+
+
+class PPHumanSeg : 
```
Propose to change the name to `MiniSupervisely`

fengyuentau commented on tools/eval/datasets/pp_humanseg.py:
```diff
@@ -0,0 +1,209 @@
+        return np.eye(max_size)[arr]
+
+
+    def calculate_area(self,pred, label, num_classes, ignore_index=255):
```
Could you provide some test data for validating the correctness of this function?

labeeb-7z replied:
For the following test : 
``` 
# tensor of size (1,x,x)
pred = np.array([
    [   [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]],
])
label = np.array([
    [   [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0]],
])

area = calculate_area(pred, label, 2)
print(f"intersect_area: {area[0]}  pred_area: {area[1]}  label_area: {area[2]}")
iou = mean_iou(*area)
print(f"IOU on classes: {iou[0]}  mean IOU: {iou[1]}")
```

Here's the output : 
![image](https://user-images.githubusercontent.com/94868003/218070001-7fe9b154-85bd-4469-822c-6d03060c5925.png)

It can be seen from the pred and label tensor that intersect_area of 0s is 20 and intersect area of 1s is 4.

For IOU : 
union for class 0 : pred + label - interset = 20 + 21 - 20 = 21
iou for class 0 : intersect/union = 20/21 = 0.9523
union for class 1 : pred + label - interset  4 + 5 - 4 = 5
iou for class 1 : intersect/union = 4/5 = 0.8

fengyuentau commented on tools/eval/datasets/pp_humanseg.py:
```diff
@@ -0,0 +1,209 @@
+        return intersect_area, pred_area, label_area
+    
+
+    def mean_iou(self,intersect_area, pred_area, label_area):
```
Could you provide some test data for validating the correctness of this function? Something like test cases with given input and expected output.

labeeb-7z replied:
please refer above

fengyuentau commented on tools/eval/eval.py:
```diff
@@ -97,6 +101,9 @@
             topic="text_recognition"),
+        pphumanseg=dict(
+            name="PPHumanSeg",
+            topic="human_segmentation"),
```
Propose to change dataset name.

fengyuentau commented on tools/eval/eval.py:
```diff
@@ -108,7 +115,6 @@ def main(args):
     model_topic = models[model_key].pop("topic")
     model_handler, _ = MODELS.get(model_name)
     model = model_handler(**models[model_key])
-
```
unnecessary removal

fengyuentau reviewed (commented):
Thank you for contribution! Please, take a look at the comments below.

WanliZhong commented:
@labeeb-7z Thanks for your contribution! Could you add the evaluation results of `fp32` and `int8` models on [PPhumanseg Readme](https://github.com/opencv/opencv_zoo/tree/master/models/human_segmentation_pphumanseg#pphumanseg)? You can add the information like the example below:

-------------------------------------------------------------------
Results of accuracy evaluation with [tools/eval](../../tools/eval).

| Models      | Mean Accuracy |
| ----------- | -------- |
| PPHumanSeg| 0.9023   |
| PPHumanSeg quant | 0.xxxx   |

\*: 'quant' stands for 'quantized'.

labeeb-7z commented:
@WanliZhong here's the output I got for quantized model : 
![image](https://user-images.githubusercontent.com/94868003/218074725-2ac87567-e5c0-427a-aef3-0606be9b70c1.png)

WanliZhong commented:
> @WanliZhong here's the output I got for quantized model : ![image](https://user-images.githubusercontent.com/94868003/218074725-2ac87567-e5c0-427a-aef3-0606be9b70c1.png)

There is too much loss of accuracy in the quantized model and we need to consider re-quantizing the model. Thank you for the evaluation script! 👍

fengyuentau commented on models/human_segmentation_pphumanseg/README.md:
```diff
@@ -22,6 +22,18 @@ python demo.py --help
+| Models             | Accuracy | 
+| ------------------ | -------------- |
+| PPHumanSeg         | 0.9023         |
+| PPHumanSeg quant   | 0.4940         |
```
Column `Accuracy` is not fully aligned.

fengyuentau commented on tools/eval/eval.py:
```diff
@@ -108,7 +119,7 @@ def main(args):
     model_handler, _ = MODELS.get(model_name)
     model = model_handler(**models[model_key])
-
+    
```
Extra whitespaces here

fengyuentau reviewed (commented):
Some minor comments below. I will have a check on the implementation of the evaluation metric later.

labeeb-7z commented:
For more reference on the evaluation metric you can have a look at the [metrics module](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.1/paddleseg/utils/metrics.py) of PaddleSeg which is used by the evaluation script of PaddleSeg I linked in PR description.

fengyuentau reviewed (commented):
I ran the official evaluation script and the mIoU of the model should be 0.8602. I found there are many differences between your implementation and the official evaluation script. Below are some differences I found.

I am not saying you have to follow the official implementation completely but just make sure it is correct. If you cannot ensure correctness, it will save a lot of time.

labeeb-7z commented:
I've moved the *_all variables outside the loop (apologies for that). As for ignoring index and adding 1, I've tried to followed the PaddleSeg implementation. And for the one_hot vectors, again I have implemented the one_hot function of PaddleSeg and followed their evaluation script.

Let me know of any other changes. The updated miou is : 
![image](https://user-images.githubusercontent.com/94868003/218458744-faa10a40-def7-4a64-97f8-88f99c2ebbb1.png)

fengyuentau commented on tools/eval/datasets/minisupervisely.py:
```diff
@@ -0,0 +1,206 @@
+        pbar = tqdm(self.image_set)
+
+        for input_image, expected_image in pbar : 
+            pbar.set_description(
+                "Evaluating {} with {} val set".format(model.name, self.name))
```
These two lines should be placed outside the for loop since the description does not need to be updated every single iteration.

labeeb-7z replied:
This was referred from [eval script of icdar](https://github.com/opencv/opencv_zoo/blob/6c68bc48c6f96042b29b3425174e431ccac38376/tools/eval/datasets/icdar.py#L37).
I've changed it for this script, should I also change it for icdar?

fengyuentau replied:
> I've changed it for this script, should I also change it for icdar?

Not in this pull request.

fengyuentau commented on tools/eval/datasets/minisupervisely.py:
```diff
@@ -0,0 +1,206 @@
+            pbar.set_description(
+                "Evaluating {} with {} val set".format(model.name, self.name))
+            
+            input_image = cv.imread(os.path.join(self.root, input_image))
+            input_image = cv.resize(input_image, (192, 192))
```
Missing two transforms:

1. to_rgb: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/transforms/transforms.py#L61-L62. Note that to_rgb is set to True by default.
2. normailze: 
    1. Config: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/contrib/HumanSeg/configs/fcn_hrnetw18_small_v1_humanseg_192x192_mini_supervisely.yml#L20-L28 
    2. Normalize transform: https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/transforms/transforms.py#L402-L444

labeeb-7z replied:
I dont see a direct transforms applied in the [infer method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/infer.py#L192) used by the [evaluation method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/val.py#L28).  The [inference script](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/legacy/contrib/HumanSeg/infer.py#L68) of the model in PaddleSeg uses it. So Im not sure if there's a need for it in this eval script

fengyuentau replied:
> mIoU has decreased to 0.3562 .

My suggestion would be doing alignment with PaddleSeg's results image by image. For example, you could add `break` right after `calculate_area` in your code and PaddleSeg, print `intersect_area, pred_area, label_area` to see whether your results are completely the same with PaddleSeg's. You will also need to ensure you are comparing the same image since PaddleSeg uses a data loader to load images which may have shuffled images.

labeeb-7z replied:
@fengyuentau The reason mIoU decresead to 0.3562 was doing normalization twice. [model.infer](https://github.com/opencv/opencv_zoo/blob/6c68bc48c6f96042b29b3425174e431ccac38376/models/human_segmentation_pphumanseg/pphumanseg.py#L44) of PPHumanSeg already does normalization before passing the input to model so we dont have to do normalization again in the evaluation script. I've reverted that back.

The reason behind why current eval script gives mIoU of 0.8798 and not the expected 0.8602 is - model outputs a 192x192 tensor(logits). Original PaddleSeg script resizes this tensor back to original image size before taking one-hot of that tensor and calculating accuracy, IoU (all calculations done on tensor of shape equal to input image).
This can be seen below : 
[infer method called by PaddleSeg evaluation script](https://github.com/PaddlePaddle/PaddleSeg/blob/13aac35b31b147d2dd9976821ade13dabebff685/paddleseg/core/infer.py#L134)
![image](https://user-images.githubusercontent.com/94868003/219100036-92a290d7-71c9-4c7b-b64a-9f2631524153.png)
Here's the output (the input image is of size 800x1313 in this case) : 
![image](https://user-images.githubusercontent.com/94868003/219100458-7d8112ef-f2cf-4ad1-a5fb-df05dcdeaa93.png)

On the other hande, current eval script resizes the expected_image (label) to 192x192 (same as output of model) and then calculates accuracy, IoU. 
This is because the infer method of PPHumanSeg does not resize the input back to its original shape and directly calculates the one-hot. And thus all the calculation (for mIoU and accuracy) is done on a 192x192 tensor and not the original image size tensor.
![image](https://user-images.githubusercontent.com/94868003/219110459-403907ce-302f-4548-a8b5-42721b94e8a5.png)

![image](https://user-images.githubusercontent.com/94868003/219110336-c33b4a66-3adc-4624-b258-1b2c5a29ea34.png)


Should I proceed with modifying the infer method of PPHumanSeg class to resize the output back to original image shape before calculating the one-hot?

fengyuentau replied:
> Should I proceed with modifying the infer method of PPHumanSeg class to resize the output back to original image shape before calculating the one-hot?

No, the idea of infer method of PPHumanSeg is assuming the input image is resized to the target size [192, 192] already. Thus, any resize operation should be done here, say right after `output_image = model.infer(input_image)   ` we `reverse_transform(output_image)`.

`to_rgb` needs to be kept or directly add this in `PPHumanSeg._preprocess`.

labeeb-7z replied:
The logits output of model.infer should be resized. model.infer returns the one-hot and not the logits.

so `output_image` is a one-hot and resizing that won't be the same as resizing the `logits` of the model.(correct me if Im wrong).

If we want to do `reverse_transform(outpu_image)` in the eval script, we should change the model.infer method to return the logits and not postprocess it to get the one-hot.

So
- resize the logits in model.infer method before calculating one-hot(postprocess)
- return the logits from model.infer and do reverse_transform in the eval scrirpt.

which one should we proceed with?

fengyuentau replied:
Okay. I propose to add resize (to target_size [192, 192]) in `_preprocess` and reverse resize (to its original size) in `_postprocess` in `PPHumanSeg` instead.

labeeb-7z replied:
I made the above changes, however the mIoU now is 0.9085 .

I started by comparing the 2 models. The [human_segmentation_pphumanseg_2021oct.onnx](https://github.com/opencv/opencv_zoo/blob/master/models/human_segmentation_pphumanseg/human_segmentation_pphumanseg_2021oct.onnx) downloaded from [this repository](https://github.com/opencv/opencv_zoo/tree/master/models/human_segmentation_pphumanseg) was evaluated on the script I've written so far. 
The [PP-HumanSegV1-Lite](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) model downloaded from PaddleSeg repo (branch release/2.7). It has mUoU of 0.8602 . It was evaluated using the [eval script](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#63-evaluation) provided by PaddleSeg

First I compared the input to both models (make sure preprocessing is done properly) : 
zoo model : 
![image](https://user-images.githubusercontent.com/94868003/219444281-01a9ef1d-24fe-445e-a4b9-a7cd2c2d1367.png)
PaddleSeg model ([infer method](https://github.com/PaddlePaddle/PaddleSeg/blob/83a419ff92c54b01257c75d75a435df6d75e7515/paddleseg/core/infer.py#L192) called by eval script): 
![image](https://user-images.githubusercontent.com/94868003/219444847-d5ffab92-40c3-450f-a13e-871fcacff3c0.png)

Both the tensors were equivalent : 
![image](https://user-images.githubusercontent.com/94868003/219444967-b4a28909-23f4-4ddc-ba45-601d44dababe.png)


Next I compared the immediate outputs of the model : 
zoo model : 
![Screenshot from 2023-02-16 23-15-02](https://user-images.githubusercontent.com/94868003/219446223-4328cf7d-a45d-4ff9-a079-598089a08960.png)
PaddleSeg model : 
![image](https://user-images.githubusercontent.com/94868003/219446323-20c0ad80-1d2b-46cc-b7b1-55effab264cd.png)

Surprisingly these 2 outputs were different!
![Screenshot from 2023-02-16 23-15-23](https://user-images.githubusercontent.com/94868003/219446404-1b4653f2-c2c8-4424-b768-14ce1905201a.png)

(I made sure both of them were using the same input image).

Most likely Im using a different model for comparision? Let me know where can I find the actual model or if Im doing something else wrong

fengyuentau replied:
> Most likely Im using a different model for comparision?

It could happen. The one here in the zoo is converted from Paddle to ONNX using [this script](https://github.com/opencv/opencv/tree/master/samples/dnn/dnn_model_runner/dnn_conversion/paddlepaddle#2-run-paddleseg-portrait-segmentation-using-opencv), and the script has been there for 2 years. Note that the model was provided in a link which is different from all other links in PaddleSeg:release/2.7. Paddle could have updated this model during this time and somehow still name it "v1".

It can be verified with using your evaluation script to evaluate the one from PaddleSeg:release/2.7. Paddle provides a conversion tool https://github.com/PaddlePaddle/Paddle2ONNX. If you can reach the same mIoU (0.8602) with your evaluation script and the model from PaddleSeg:release/2.7, that's a success.

labeeb-7z replied:
I tried to convert the models from [PaddleSeg](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) into onnx format.

the PP-HumanSegV1-Lite (checkpoint) gives the following error
![image](https://user-images.githubusercontent.com/94868003/219581274-061ebb1f-cb9f-42c9-991c-47846c11fa38.png)

I tried other models (v2,mobile) from there, none of the checkpoint models were converted to onnx (same error).

then I tried the Inference Model(softmax) of v1. It was converted to onnx successfully.
However, `cv.dnn.readNet()` wasn't able to read it : 
![image](https://user-images.githubusercontent.com/94868003/219590152-4d9774d4-5291-4d91-923b-d4ebbff1e7b1.png)

The readNet function gave the same error for Mobile inference model(softmax).

At this point, PaddleSeg doesn't provide the model currently in this repository, and the models they provide either don't convert to onnx or are not readable by readNet().

What should be the next steps?

fengyuentau replied:
> However, cv.dnn.readNet() wasn't able to read it 

The error was due to lack of support of dynamic input shape (dynamic H or W) I believe. You can use https://github.com/daquexian/onnx-simplifier to simplify it with a fixed input shape, such as [1, 3, 192, 192]. After that, OpenCV DNN should be able to load and infer the model.

fengyuentau replied:
> However, cv.dnn.readNet() wasn't able to read it 

The error was due to lack of support of dynamic input shape (dynamic H or W) I believe. You can use https://github.com/daquexian/onnx-simplifier to simplify it with a fixed input shape, such as [1, 3, 192, 192]. After that, OpenCV DNN should be able to load and infer the model.

labeeb-7z replied:
I converted the v1 model using the command 
`onnxsim seg.onnx v1sim.onnx --overwrite-input-shape 1,3,192,192`

but now the onnx model is corrupted in some way : 
![Screenshot from 2023-02-17 15-32-43](https://user-images.githubusercontent.com/94868003/219615242-a7da1fef-bbad-4142-bf74-47413e8eb9fc.png)

fengyuentau replied:
> I tried the Inference Model(softmax) of v1. It was converted to onnx successfully.

Did it really work? I downloaded the same model and use the same command as yours but the produced model has 0 Bytes.

labeeb-7z replied:
On [PaddleSeg](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.7/contrib/PP-HumanSeg#42-general-human-segmentation-models) for PP-HumanSegV1-Lite, the model downloaded with `checkpoint` link gives the error and produces a model of 0 Bytes.

But the model downloaded with `Inference Model (Softmax)` link is converted to onnx succesfully : 
![image](https://user-images.githubusercontent.com/94868003/219874704-3bea8e34-59eb-429d-ad46-3f452839b59b.png)
![image](https://user-images.githubusercontent.com/94868003/219874720-0e0dd8d2-5066-4f1c-93d8-75e54307538f.png)

I simplified the onnx  models produced by `Inference Model (Softmax)` and than tried the eval script on them.
These models were not read by `cv.dnn.readNet` which I shared above

fengyuentau replied:
Thank you for pointing out the correct model and command. The conversion was done successfully and the simplifier worked as well. Latest OpenCV (4.7) can load and infer the model without any issues. See my screen shot below. You could have a try with my model: [humanseg.sim.onnx.zip](https://github.com/opencv/opencv_zoo/files/10779034/humanseg.sim.onnx.zip)


![image](https://user-images.githubusercontent.com/17219438/219999042-0b46ac98-7e0c-4263-9524-2915ea7a531a.png)

labeeb-7z replied:
Thanks, the error was fixed by removing the `self.outputNames` argument from `model.forward()`.

The eval script achieved the expected mIoU of 0.8602 on the model you shared above  (humanseg.sim.onnx) : 
![image](https://user-images.githubusercontent.com/94868003/220078200-d04dda58-3564-4f15-9bc6-267ca5b0dcda.png)

This matches the results from the PaddleSeg script : 
![image](https://user-images.githubusercontent.com/94868003/220078958-25458ce2-e9fa-4df9-92d6-3cd00d1addf8.png)
 
With the same script, for the model present in the zoo, the mIoU is 0.8996 : 
![image](https://user-images.githubusercontent.com/94868003/220079082-b248a1be-5256-4e00-a611-42928890d144.png)

fengyuentau replied:
👍

fengyuentau reviewed (commented):
Thank you for the update! Lets try to reproduce the exact mIoU number.

fengyuentau commented on tools/eval/datasets/minisupervisely.py:
```diff
@@ -0,0 +1,201 @@
+        print("Mean Accuracy : ", self.acc)
+
+    
+    def one_hot(self, arr, max_size) :
+        return np.eye(max_size)[arr]
+
+
```
Looks like it is not used anymore. Suggest to remove this.

fengyuentau commented on tools/eval/datasets/__init__.py:
```diff
@@ -20,4 +21,5 @@ def register(self, item):
-DATASETS.register(IIIT5K)
\ No newline at end of file
+DATASETS.register(IIIT5K)
+DATASETS.register(MiniSupervisely)
```
nit: put a newline at the end of the file.

fengyuentau commented on tools/eval/README.md:
````diff
@@ -190,4 +191,24 @@ Run evaluation with the following command:
+```
+
+
+## mini_supervisely
````
-> Mini Supervisely

fengyuentau commented on tools/eval/README.md:
```diff
@@ -190,4 +191,24 @@ Run evaluation with the following command:
+
+### Evaluation
+
+Run evaluation with the following command
```
add a colon at the end of this line

fengyuentau commented on tools/eval/README.md:
````diff
@@ -190,4 +191,24 @@ Run evaluation with the following command:
+python eval.py -m pphumanseg -d mini_supervisely -dr /path/to/pphumanseg
+```
+
+Run evaluation on quantized model with the following command
````
add a colon at the end of this line

fengyuentau commented on tools/eval/datasets/minisupervisely.py:
```diff
@@ -0,0 +1,196 @@
+from tqdm import tqdm
+
+
+class MiniSupervisely : 
```
Could you add a comment as reference to the original evaluation method in PaddleSeg? Something like this:

```.diff
class MiniSupervisely : 
+    '''
+       Refer to https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.7/paddleseg/core/val.py
+       for official evaluation implementation.
+    '''
```

fengyuentau reviewed (commented):
Overall looks good 👍 We can merge this pull request after the comments below are resolved.

labeeb-7z commented:
@fengyuentau I've made the suggested changes, Thanks!

fengyuentau reviewed (approved):
Thank you for contribution! 👍
